torch/_inductor/scheduler.py

```diff
@@ -2978,8 +2978,20 @@ def codegen(self) -> None:
         with dynamo_timed("Scheduler.codegen"):
             return self._codegen()
 
+    def get_training_phase(self) -> str:
+        if V.graph.is_inference:
+            return "inference"
+        if V.graph.is_backward:
+            return "backward"
+        return "forward"
+
     def _codegen(self) -> None:
+        phase = self.get_training_phase()
+        if config.annotate_training:
+            V.graph.wrapper_code.writeline(f"training_annotation = nvtx.device_range_start('{phase}')")
```

Comment:
Let's try to keep codegen aspects out of scheduler and keep high level scheduler logic lean. This annotation can go here: https://github.com/pytorch/pytorch/blob/1da3a049dac3c78554506d5ef9ede55b7c2b774d/torch/_inductor/codegen/wrapper.py#L670

Similarly, the end can go in `_generate`.